proposals/assert/design_by_contract.txt

```diff
@@ -0,0 +1,247 @@
+
+  SUBROUTINE PROBLEM_CALL ( ..., A, ... )
+    ...
+    TYPE(EXAMPLE), POINTER :: A
+    ...
+    CALL PROBLEM( A )
+    ASSERT( ASSOCIATED( A ) )
+    ...
+    RETURN
+  END SUBROUTINE PROBLEM_CALL
+
+
+4. Prior art
+
+The term design by contract as applied to programming languages has
+its origins in the language Eiffel.[4,5] Since then, in addition to
+C++ it has been adopted by Ada,[6] D,[7] and Fortress. The C standard
+library comes with the assert macro so C can be said to partially
+support DbC. To our knowledge no Fortran compiler has implemented
+anything to support DbC. Many users implement their own equivalent of
+an ASSERT subroutine or macro so there is some demand for these
+capabilities.
+
+
+5. Specification
+
+All three statements will have similar behavior. All three will take
+as their arguments a list of scalar logical expressions, that if all
+.TRUE. does nothing, and if one is .FALSE. writes output to the
+ERROR_UNIT of ISO_FORTRAN_ENV and stops as if ERROR STOP was
+invoked. The writes will all indicate the location of the executed
+statement: either the procedure name with the name of the enclosing
+module, or the name of the file including the statement and the line
+number of the statement, or both. All three will write out the text of
+the logical expression that failed, with a prefix. The prefix will
+differ between the three statements. Example prefixes for the REQUIRES
+statement is "PRE-CONDITION FAILED: ", for the ASSERTS statement is
+"ASSERTION FAILED: ". and for the ENSURES statement is "POST-CONDITION
+FAILED: ".
+
+The semantics of all three statements are similar. They all should
+respond the same to compiler flags that invoke different behavior in
+development versus in deployment. In development, they each perform a
+logical test with the .FALSE. branch writing to ERROR_UNIT, and then
+stopping with an stop code, and the other branch a null operation
+that continues processing. In deployment they are as if they were
+commented out. These semantics are similar to a logical test with one
+branch invoking ERROR STOP and the other branch continuing
+processing. They should have restrictions similar to those of ERROR
+STOP, being allowed in PURE procedures, but not in ELEMENTAL
+procedures.
+
+
+4. Syntactic options
+
+The syntax for the statements depend on whether they should be
+considered executable statements, and should go in the executable
+part, or attributes of the sub-programs, and should go in the
+specification part. It is difficult to think of ASSERTS as any thing
+other than an executable statement. For ASSERTS I think the following
+syntax is appropriate 
+
+  asserts-stmt is ASSERTS( assertion-list )
+
+where
+
+  assertion is logical-expr
+
+REQUIRE and ENSURES can be thought of as either sub-program attributes
+or executable statements. If treated as sub-program attributes then
+the syntax could be something like
+
+  requires-stmt is EEQUIRES :: requires-expr-list
```

Comment:
```suggestion
  requires-stmt is REQUIRES :: requires-expr-list
```

Reply:
@wclodius2 If you like the edits, you can accept them by clicking on "Commit suggestion".